exec/java-exec/src/main/java/org/apache/drill/exec/expr/fn/impl/DateTypeFunctions.java

```diff
@@ -470,43 +500,73 @@ public void eval() {
+                // negative period and positive duration
+                period = java.time.Period.between(from.toLocalDate(), to.toLocalDate().plusDays(1));
+                duration = duration.minusDays(1);
+            } else if (from.isBefore(to) && duration.compareTo(java.time.Duration.ZERO) < 0) {
+                // positive period and negative duration
+                period = java.time.Period.between(from.toLocalDate(), to.toLocalDate().minusDays(1));
+                duration = duration.plusDays(1);
+            } else {
+                period = java.time.Period.between(from.toLocalDate(), to.toLocalDate());
+            }
+
+            out.months = (int) period.toTotalMonths();
+            out.days = period.getDays();
+            out.milliseconds = (int) duration.toMillis();
         }
     }
 
     @FunctionTemplate(name = "age", scope = FunctionTemplate.FunctionScope.SIMPLE, nulls = NullHandling.NULL_IF_NULL)
     public static class AgeDate2Function implements DrillSimpleFunc {
         @Param DateHolder right;
-        @Workspace long queryStartDate;
+        @Workspace java.time.OffsetDateTime queryStartDate;
         @Output IntervalHolder out;
         @Inject ContextInformation contextInfo;
 
         @Override
         public void setup() {
             int timeZoneIndex = contextInfo.getRootFragmentTimeZone();
-            org.joda.time.DateTimeZone timeZone = org.joda.time.DateTimeZone.forID(org.apache.drill.exec.expr.fn.impl.DateUtility.getTimeZone(timeZoneIndex));
-            org.joda.time.DateTime now = new org.joda.time.DateTime(contextInfo.getQueryStartTime(), timeZone);
-            queryStartDate = (new org.joda.time.DateMidnight(now.getYear(), now.getMonthOfYear(), now.getDayOfMonth(), timeZone)).getMillis();
+            java.time.ZoneId zoneId = java.time.ZoneId.of(org.apache.drill.exec.expr.fn.impl.DateUtility.getTimeZone(timeZoneIndex));
+            java.time.ZonedDateTime dt = java.time.Instant.ofEpochMilli(contextInfo.getQueryStartTime()).atZone(zoneId);
+            queryStartDate = java.time.OffsetDateTime.of(dt.toLocalDate(), java.time.LocalTime.MIDNIGHT, java.time.ZoneOffset.UTC);
         }
 
         @Override
         public void eval() {
-            long diff = queryStartDate - right.value;
-            long days = diff / org.apache.drill.exec.vector.DateUtilities.daysToStandardMillis;
-            out.months = (int) (days / org.apache.drill.exec.vector.DateUtilities.monthToStandardDays);
-            out.days = (int) (days % org.apache.drill.exec.vector.DateUtilities.monthToStandardDays);
-            out.milliseconds = (int) (diff % org.apache.drill.exec.vector.DateUtilities.daysToStandardMillis);
+            java.time.OffsetDateTime dt = java.time.Instant.ofEpochMilli(right.value).atOffset(java.time.ZoneOffset.UTC);
+            java.time.Duration duration = java.time.Duration.between(dt.toLocalTime(), queryStartDate.toLocalTime());
+            java.time.Period period;
```

Comment:
Shame we have to do memory allocations in the per-column inner loop. Drill goes to heroic lengths to avoid such allocations. Is there some "modern" library which can do these calcs without creating garbage? Or, will Drill's (or the JVM's) optimizer rewrite the code to avoid garbage?

Reply:
@paul-rogers 
Since I still have to rewrite the implementation taking into account the comments below, I can try to write an alternative implementation using the source code of Java Time

Reply:
@oleg-zinovev, please don't go to heroic lengths to rewrite the date functions. Date code is complex, we do want to use an existing library. The question is more along the lines of, "so the Joda and Java techniques work, but create many temporary objects in the inner per-row loop. Is there some other *existing* library we can use?" If there isn't, then we just live with the performance. After all `AGE()` is a very old-school function as it is not stable in a distributed environment. It made more sense back when people used just `DATE` types on a single-processor machine. So, the function probably won't be used much. That's why we found the issue now, after many years.

Reply:
I rewrote the function implementation using an idea from the Joda time BasePeriod